numpy/testing/tests/test_utils.py

```diff
@@ -1114,18 +1114,26 @@ def test_simple(self):
                           lambda: assert_string_equal("foo", "hello"))
 
 
-def assert_warn_len_equal(mod, n_in_context, py3_n_in_context=None):
+def assert_warn_len_equal(mod, n_in_context, py34=None, py37=None):
     mod_warns = mod.__warningregistry__
+    num_warns = len(mod_warns)
     # Python 3.4 appears to clear any pre-existing warnings of the same type,
     # when raising warnings inside a catch_warnings block. So, there is a
     # warning generated by the tests within the context manager, but no
     # previous warnings.
     if 'version' in mod_warns:
-        if py3_n_in_context is None:
-            py3_n_in_context = n_in_context
-        assert_equal(len(mod_warns) - 1, py3_n_in_context)
-    else:
-        assert_equal(len(mod_warns), n_in_context)
+        # Python 3 adds a 'version' entry to the registry,
+        # do not count it.
+        num_warns -= 1
+
+        # Behavior of warnings is Python version dependent. Adjust the
+        # expected result to compensate. In particular, Python 3.7 does
+        # not make an entry for ignored warnings.
+        if py34 is not None and sys.version_info[:2] < (3, 7):
+            n_in_context = py34
+        if py37 is not None and sys.version_info[:2] >= (3, 7):
+            n_in_context = py37
```

Comment:
The mutual exclusion is a little clearer as:

```
if sys.version_info[:2] < (3, 7):
    if py34 is not None:
        n_in_context = py34
else:
    if py37 is not None:
        n_in_context = py37
```
but generally much clearer than what was here before

Reply:
I deliberately chose not to do it that way because:

* This may not be the last change
* I liked having all the versions laid out

Reply:
I don't know what you mean by either of those points:

* Are you suggesting py3.8 may change it again? It's easier to chain an`if` onto my version than onto what this patch has
* Doesn't my version lay out the versions more clearly? The python versions aren't tangled between branches this way